File: C++/Bryann Valderrama/Matemática/josephus.cpp
#include <cstdlib>
#include <bits/stdc++.h>
using namespace std;

int next(int a, int k, int d[], int n)
{
    int j = a - 1; //posicion anterior de "a"
    for (int i = 0; i < k;)
    {
        j = (j + 1) % n;
        if (d[j] != -1)
        {
            i++;
        }
    }
    cout << "Next = " << j << endl;
    return j;
}

int josephus(int n, int k)
{
    int d[n + 1];
    d[0] = -1; //Cuando comienza desde 1
    for (int i = 1; i <= n; i++)
    {
        d[i] = i;
    }
    int indice = n;
    int a = 0; //calculo de la posicion siguiente a saltar
    while (indice != 1)
    {
        indice--;
        a = next(a, k, d, n + 1);
        d[a] = -1;
        a++;
    }
    for (indice = 0; d[indice] == -1; indice++)
        ;
    return d[indice];
}

int josephusModular(int n, int k)
{
    int f = 0;
    for (int i = 1; i <= n; i++)
    {
        f = (f + k) % i;
    }
    return f + 1;
}

int main(int argc, char const *argv[])
{
    ios_base::sync_with_stdio(false);
    cout.tie(NULL);
    cout << josephusModular(6, 2) << endl;
    cout << josephus(6, 2) << endl;
    return 0;
}

File: Codigos estudiantes por lenguaje/PY/Bryann Valderrama/Matemática/josephus.py
from sys import stdout
wr = stdout.write


def next_(a, k, d):
    j = a - 1
    i = 0
    while i < k:
        j = (j+1) % len(d)
        if d[j] != -1:
            i += 1
    wr(f'Next = {j}\n')
    return j


def josephus(n, k):
    d = [-1 for x in range(n+1)]
    d[0] = -1
    for i in range(1, n+1):
        d[i] = i
    i = n
    a = 0
    while i != 1:
        i -= 1
        a = next_(a, k, d)
        d[a] = -1
        a += 1
    i = 0
    while d[i] == -1:
        i += 1
    return d[i]


def josephusModular(n, k):
    f = 0
    for i in range(1, n+1):
        f = (f+k) % i
    return f + 1


wr(f'{josephusModular(6,2)}\n')
wr(f'{josephus(6,2)}')
